lib/components/cluster.js

```diff
@@ -3,11 +3,116 @@
 import React from 'react';
 
 /**
+ * Types.
+ */
+
+type Props = {
+  children: Array<ReactElement>,
+  className: ?string,
+  height: number,
+  rowHeight: number,
+};
+
+type State = {
+  index: number,
+};
+
+/**
+ * Defaults and options.
+ */
+
+const INITIAL_STATE =  {
+  index: 0,
+};
+
+const RANGE_OFFSET = 5;
+
+/**
  * Cluster component.
+ *
+ * Displays large sets of data in a scroll container.
  */
 
-export default function Cluster(): ReactElement {
-  return (
-    <span>hello world</span>
-  );
+export default class Cluster extends React.Component {
+  props: Props;
+  state: State = INITIAL_STATE;
+
+  shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
+    return this.props.children !== nextProps.children || this.state.index !== nextState.index;
+  }
+
+  componentDidMount(): void {
+    this.cluster.addEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  componentWillUnmount(): void {
+    this.cluster.removeEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  setIndex(): void {
+    this.setState({ index: this.findIndex() });
+  }
+
+  render(): ReactElement {
+    const {children, className, height, rowHeight} = this.props;
+
+    const rows = toClusterRows({
+      children,
+      height: rowHeight,
+      range: this.findRange(),
+    });
+
+    return (
+      <div className={className} ref={(ref) => this.cluster = ref} style={{ height, overflow: 'scroll' }}>
+        {rows}
+      </div>
+    );
+  }
+
+  findIndex(): number {
+    return Math.floor(this.cluster.scrollTop / this.props.rowHeight);
+  }
+
+  findRange() : Array<number> {
+    const {height, rowHeight} = this.props;
+    const {index} = this.state;
+
+    const start = index - RANGE_OFFSET;
+    const end = index + Math.ceil(height / rowHeight) + RANGE_OFFSET;
+
+    return range(start, end);
+  }
+}
+
+/**
+ * Return an array of cluster row indexes.
+ */
+
+function range(start: number, end: number): Array<number> {
+  const length = end - start;
+
+  let index = -1;
+  let result = Array(length);
+
+  while (++index < length) {
+    result[index] = start;
+    start += 1;
+  }
+
+  return result;
+}
```

Comment:
this feels overly complex

eg 

- why result[index] vs array.push
- why not using `_.range` 

but tbh range feels like an overkill

while i would do this the same way in ruby it's too much for here
you are doing this to be able to do `.includes()` below


i would do 
```
const [firstVisibleRow, lastVisibleRow] = this.findRange();
(index > firstVisibleRow && index < lastVisibleRow) ? child : null
```